File: apps/opensible-server/services/approval_service.py
"""Approval workflow (Fase 2 — UC 50/68/72)."""
from __future__ import annotations

import json
import time
import uuid
from pathlib import Path
from typing import Any, Dict, List, Optional


def _store_path() -> Path:
    try:
        import app as _app
        return Path(getattr(_app, "DATA_DIR", "data")) / "approvals.json"
    except Exception:
        return Path("data") / "approvals.json"


def _load() -> List[Dict[str, Any]]:
    try:
        p = _store_path()
        if p.exists():
            d = json.loads(p.read_text(encoding="utf-8"))
            if isinstance(d, list):
                return d
    except Exception:
        pass
    return []


def _save(records: List[Dict[str, Any]]) -> None:
    p = _store_path()
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(records, indent=2), encoding="utf-8")


def trigger_approval_retest(project_id: Optional[str], stack: str, approval_id: Optional[str] = None) -> Optional[str]:
    """Trigger re-test for approval request (UC191)."""
    try:
        from services import test_cases
        if hasattr(test_cases, "trigger_approval_retest"):
            return test_cases.trigger_approval_retest(project_id, stack, approval_id)
        if hasattr(test_cases, "run_all_tests"):
            res = test_cases.run_all_tests(project_id=project_id, stack=stack)
            if res and res.get("results"):
                return res["results"][0].get("run_id") or str(uuid.uuid4())
            return str(uuid.uuid4())
        if hasattr(test_cases, "run_batch_tests"):
            res = test_cases.run_batch_tests(project_id=project_id, stack=stack)
            if res and res.get("results"):
                return res["results"][0].get("run_id") or str(uuid.uuid4())
            return str(uuid.uuid4())
    except Exception:
        pass
    return None


def is_approval_expired(approval_dict: Dict[str, Any]) -> bool:
    """Check if an approval request has expired (UC615)."""
    if not approval_dict:
        return False
    expires_at = approval_dict.get("expires_at")
    if expires_at and approval_dict.get("status") == "pending":
        return time.time() > float(expires_at)
    return False


def create_approval(stack: str, project_id: str, action: str,
                    requested_by: str = "", note: str = "",
                    ttl_seconds: int = 86400) -> Dict[str, Any]:
    approval_id = str(uuid.uuid4())
    retest_run_id = None
    try:
        retest_run_id = trigger_approval_retest(project_id=project_id, stack=stack, approval_id=approval_id)
    except Exception:
        pass

    now = time.time()
    rec = {
        "id": approval_id,
        "stack": stack,
        "project_id": project_id,
        "action": action,
        "status": "pending",
        "requested_by": requested_by,
        "note": note,
        "created_at": now,
        "expires_at": now + max(60, int(ttl_seconds)),
        "decided_at": None,
        "decided_by": None,
        "retest_run_id": retest_run_id,
    }
    records = _load()
    records.append(rec)
    _save(records)
    return rec


def decide(approval_id: str, status: str, decided_by: str = "", reason: str = "") -> Optional[Dict[str, Any]]:
    if status == "rejected" and not (reason and str(reason).strip()):
        raise ValueError("rejection reason is mandatory")

    records = _load()
    for r in records:
        if r.get("id") == approval_id:
            # Check for TTL expiry
            if is_approval_expired(r):
                r["status"] = "expired"
                r["decided_at"] = time.time()
                r["decided_by"] = "system"
                _save(records)
                return r

            r["status"] = status
            r["decided_at"] = time.time()
            r["decided_by"] = decided_by
            if reason:
                r["rejection_reason"] = str(reason).strip()
            _save(records)
            if status == "approved" and r.get("action") == "apply":
                # Auto-apply after review (UC 51).
                try:
                    from services.cloud_provisioning import _create_execution
                    _create_execution(r.get("project_id"), r.get("stack"), "apply",
                                      triggered_by=f"approval:{approval_id}")
                except Exception:
                    pass
            return r
    return None


def get_approval(approval_id: str) -> Optional[Dict[str, Any]]:
    """Retrieve an approval by ID."""
    records = _load()
    for r in records:
        if r.get("id") == approval_id:
            return r
    return None


def approve_approval(approval_id: str, decided_by: str = "", note: str = "") -> Optional[Dict[str, Any]]:
    """Approve an approval request (UC617)."""
    return decide(approval_id, "approved", decided_by=decided_by, reason=note)


def reject_approval(approval_id: str, rejected_by: str = "", reason: str = "") -> Optional[Dict[str, Any]]:
    """Reject an approval request with mandatory reason (UC616)."""
    return decide(approval_id, "rejected", decided_by=rejected_by, reason=reason)


request_approval = create_approval




def list_approvals(project_id: Optional[str] = None, status: Optional[str] = None) -> List[Dict[str, Any]]:
    records = _load()
    changed = False
    for r in records:
        if is_approval_expired(r):
            r["status"] = "expired"
            changed = True
    if changed:
        _save(records)

    out = []
    for r in records:
        if project_id and r.get("project_id") != project_id:
            continue
        if status and r.get("status") != status:
            continue
        out.append(r)
    out.sort(key=lambda x: x.get("created_at") or 0, reverse=True)
    return out



def has_approved(stack: str, project_id: str, action: str) -> bool:
    for r in list_approvals(project_id=project_id):
        if (r.get("stack") == stack and r.get("action") == action
                and r.get("status") == "approved"):
            return True
    return False


def latest_pending(stack: str, project_id: str, action: str) -> Optional[Dict[str, Any]]:
    for r in list_approvals(project_id=project_id):
        if (r.get("stack") == stack and r.get("action") == action
                and r.get("status") == "pending"):
            return r
    return None


def should_skip_approval(stack: str, project_id: str, action: str = "apply", env: str = "",
                         org_id: Optional[str] = None) -> bool:
    """Evaluate whether approval gate can be skipped based on feature flags (UC128)."""
    try:
        from services.feature_flag_registry import evaluate
    except Exception:
        return False

    candidate_keys = [
        f"approval.skip.{action}",
        f"approval.{action}.skip",
        "approval.skip",
        "approval.auto_approve",
        f"stack.{stack}.skip_approval",
        f"approval.stack.{stack}.skip",
    ]

    for key in candidate_keys:
        try:
            res = evaluate(key, env=env or "prod", project_id=project_id, org_id=org_id)
            if res and res.get("enabled"):
                return True
        except Exception:
            continue

    return False


# ---------------------------------------------------------------------------
# UC362: Multi-step Approval Workflow Chain
# ---------------------------------------------------------------------------

def create_approval_chain(
    stack: str,
    project_id: str,
    action: str,
    steps: List[str],
    requested_by: str = "",
    note: str = "",
) -> Dict[str, Any]:
    """Create a sequential multi-step approval chain (UC362)."""
    approval_id = str(uuid.uuid4())
    clean_steps = [str(s).strip() for s in (steps or ["tech-lead", "devops"]) if str(s).strip()]
    if not clean_steps:
        clean_steps = ["tech-lead", "devops"]

    chain_state = [
        {
            "step": s,
            "status": "pending",
            "approver": None,
            "approved_at": None,
        }
        for s in clean_steps
    ]

    retest_run_id = None
    try:
        retest_run_id = trigger_approval_retest(project_id=project_id, stack=stack, approval_id=approval_id)
    except Exception:
        pass

    rec = {
        "id": approval_id,
        "stack": stack,
        "project_id": project_id,
        "action": action,
        "status": "pending",
        "is_chain": True,
        "steps": chain_state,
        "current_step_index": 0,
        "current_step": clean_steps[0],
        "requested_by": requested_by,
        "note": note,
        "created_at": time.time(),
        "decided_at": None,
        "decided_by": None,
        "retest_run_id": retest_run_id,
    }
    records = _load()
    records.append(rec)
    _save(records)
    return rec


def approve_chain_step(
    approval_id: str,
    step_name: Optional[str] = None,
    approver: str = "approver",
    decision: str = "approved",
) -> Dict[str, Any]:
    """Approve or reject a specific step in a multi-step approval chain (UC362)."""
    records = _load()
    target = None
    for r in records:
        if r.get("id") == approval_id:
            target = r
            break

    if not target:
        raise ValueError("approval not found")

    if not target.get("is_chain"):
        # Single step approval fallback
        return decide(approval_id, decision, decided_by=approver) or target

    if target.get("status") in ("approved", "rejected"):
        return target

    steps = target.get("steps") or []
    idx = target.get("current_step_index", 0)

    if decision == "rejected":
        target["status"] = "rejected"
        target["decided_at"] = time.time()
        target["decided_by"] = approver
        if idx < len(steps):
            steps[idx]["status"] = "rejected"
            steps[idx]["approver"] = approver
            steps[idx]["approved_at"] = time.time()
        _save(records)
        return target

    # Find step to approve
    if idx < len(steps):
        step_rec = steps[idx]
        if step_name and step_rec.get("step") != step_name:
            raise ValueError(f"Expected approval for step '{step_rec.get('step')}', got '{step_name}'")

        step_rec["status"] = "approved"
        step_rec["approver"] = approver
        step_rec["approved_at"] = time.time()

        next_idx = idx + 1
        if next_idx >= len(steps):
            # All steps completed!
            target["status"] = "approved"
            target["current_step_index"] = next_idx
            target["current_step"] = None
            target["decided_at"] = time.time()
            target["decided_by"] = approver
            _save(records)

            if target.get("action") == "apply":
                try:
                    from services.cloud_provisioning import _create_execution
                    _create_execution(target.get("project_id"), target.get("stack"), "apply",
                                      triggered_by=f"approval:{approval_id}")
                except Exception:
                    pass
        else:
            target["current_step_index"] = next_idx
            target["current_step"] = steps[next_idx].get("step")
            _save(records)

    return target


# ---------------------------------------------------------------------------
# UC614: Multi-party Approval Quorum Workflow
# ---------------------------------------------------------------------------

def create_quorum_approval(
    stack: str,
    project_id: str,
    action: str = "apply",
    min_approvals: int = 2,
    requested_by: str = "",
    note: str = "",
    ttl_seconds: int = 86400,
) -> Dict[str, Any]:
    """Create an approval request requiring quorum of multiple signatures (UC614)."""
    approval_id = str(uuid.uuid4())
    now = time.time()
    rec = {
        "id": approval_id,
        "stack": stack,
        "project_id": project_id,
        "action": action,
        "status": "pending",
        "is_quorum": True,
        "min_approvals": max(1, int(min_approvals)),
        "signatures": [],
        "requested_by": requested_by,
        "note": note,
        "created_at": now,
        "expires_at": now + ttl_seconds,
        "decided_at": None,
        "decided_by": None,
    }
    records = _load()
    records.append(rec)
    _save(records)
    return rec


def record_approval_signature(
    approval_id: str,
    approver: str,
    approver_name: str = "",
) -> Dict[str, Any]:
    """Record an approver's signature towards the quorum (UC614)."""
    records = _load()
    target = next((r for r in records if r.get("id") == approval_id), None)
    if not target:
        raise ValueError("approval request not found")

    if target.get("status") != "pending":
        return target

    now = time.time()
    signatures = list(target.get("signatures") or [])
    # Check if already signed by this user
    if any(s.get("approver") == approver for s in signatures):
        return target

    signatures.append({
        "approver": approver,
        "approver_name": approver_name or approver,
        "signed_at": now,
    })
    target["signatures"] = signatures

    min_required = int(target.get("min_approvals", 1))
    if len(signatures) >= min_required:
        target["status"] = "approved"
        target["decided_at"] = now
        target["decided_by"] = approver
        if target.get("action") == "apply":
            try:
                from services.cloud_provisioning import _create_execution
                _create_execution(target.get("project_id"), target.get("stack"), "apply",
                                  triggered_by=f"approval:{approval_id}")
            except Exception:
                pass

    _save(records)
    return target


def is_quorum_reached(approval_id: str) -> bool:
    """Check if quorum threshold of approvals has been reached (UC614)."""
    records = _load()
    target = next((r for r in records if r.get("id") == approval_id), None)
    if not target:
        return False
    if target.get("status") == "approved":
        return True
    signatures = target.get("signatures") or []
    min_required = int(target.get("min_approvals", 1))
    return len(signatures) >= min_required

File: apps/server_elixir/lib/radas_ai/approval_service.ex
defmodule RadasAI.ApprovalService do
  @moduledoc """
  Port of `services/approval_service.py` (Fase 2 — UC 50/68/72 + UC 615/616/
  617/128). Approvals live in `DATA_DIR/approvals.json` (shared store shape
  with Flask). `decide/4` auto-queues an `apply` run after approval (UC 51)
  via `RadasAI.CloudStacks.create_execution/4`.
  """

  alias RadasAI.CloudStacks
  alias RadasAI.Flags
  alias RadasAI.ProjectPaths

  @actions ["apply", "destroy", "plan"]

  defp store_path, do: Path.join([ProjectPaths.data_dir(), "approvals.json"])

  defp load do
    case File.read(store_path()) do
      {:ok, binary} ->
        case Jason.decode(binary) do
          {:ok, recs} when is_list(recs) -> Enum.filter(recs, &is_map/1)
          _ -> []
        end

      _ ->
        []
    end
  end

  defp save(records) do
    path = store_path()
    File.mkdir_p!(Path.dirname(path))
    File.write!(path, Jason.encode!(records, pretty: true))
    records
  end

  defp expired?(rec) do
    expires_at = rec["expires_at"]
    rec["status"] == "pending" and is_number(expires_at) and now() > expires_at
  end

  defp now, do: System.system_time(:second)

  @doc "Create a pending approval (TTL 24h default). Raises on empty stack."
  @spec create_approval(String.t(), String.t(), String.t(), keyword()) :: map()
  def create_approval(stack, project_id, action, opts \\ []) do
    requested_by = Keyword.get(opts, :requested_by, "")
    note = Keyword.get(opts, :note, "")
    ttl = Keyword.get(opts, :ttl_seconds, 86400)
    now = now()

    rec = %{
      "id" => Ecto.UUID.generate(),
      "stack" => stack,
      "project_id" => project_id,
      "action" => action,
      "status" => "pending",
      "requested_by" => requested_by,
      "note" => note,
      "created_at" => now,
      "expires_at" => now + max(60, trunc(ttl)),
      "decided_at" => nil,
      "decided_by" => nil,
      "retest_run_id" => nil
    }

    save(load() ++ [rec])
    rec
  end

  @doc """
  Decide one approval. Rejections require a reason. Expired pending records
  flip to `expired` instead (UC615). Approval of `apply` auto-queues the run
  (UC 51). Returns nil when the id is unknown.
  """
  @spec decide(String.t(), String.t(), keyword()) :: map() | nil
  def decide(approval_id, status, opts \\ []) do
    decided_by = Keyword.get(opts, :decided_by, "")
    reason = Keyword.get(opts, :reason, "")

    if status == "rejected" and String.trim(reason) == "" do
      raise ArgumentError, message: "rejection reason is mandatory"
    end

    records = load()

    {records, result} =
      Enum.map_reduce(records, nil, fn r, acc ->
        if r["id"] != approval_id do
          {r, acc}
        else
          if expired?(r) do
            r =
              r
              |> Map.put("status", "expired")
              |> Map.put("decided_at", now())
              |> Map.put("decided_by", "system")

            {r, r}
          else
            r =
              r
              |> Map.put("status", status)
              |> Map.put("decided_at", now())
              |> Map.put("decided_by", decided_by)
              |> then(fn r -> if reason != "", do: Map.put(r, "rejection_reason", String.trim(reason)), else: r end)

            if status == "approved" and r["action"] == "apply" do
              try do
                CloudStacks.create_execution(r["project_id"], r["stack"], "apply",
                  triggered_by: "approval:#{approval_id}"
                )
              rescue
                _ -> nil
              end
            end

            {r, r}
          end
        end
      end)

    if result do
      save(records)
    end

    result
  end

  @doc "One approval by id, or nil."
  @spec get_approval(String.t()) :: map() | nil
  def get_approval(approval_id), do: Enum.find(load(), &(&1["id"] == approval_id))

  @doc "Newest-first listing with optional project/status filters; lazily expires."
  @spec list_approvals(String.t() | nil, String.t() | nil) :: [map()]
  def list_approvals(project_id \\ nil, status \\ nil) do
    records = load()

    {records, changed} =
      Enum.map_reduce(records, false, fn r, changed ->
        if expired?(r) do
          {Map.merge(r, %{"status" => "expired", "decided_at" => now(), "decided_by" => "system"}), true}
        else
          {r, changed}
        end
      end)

    if changed, do: save(records)

    records
    |> Enum.filter(&(project_id in [nil, ""] or &1["project_id"] == project_id))
    |> Enum.filter(&(status in [nil, ""] or &1["status"] == status))
    |> Enum.sort_by(&(&1["created_at"] || 0), :desc)
  end

  @doc "Any approved record for (stack, action)?"
  @spec has_approved?(String.t(), String.t(), String.t()) :: boolean()
  def has_approved?(stack, project_id, action) do
    Enum.any?(list_approvals(project_id), &(&1["stack"] == stack and &1["action"] == action and &1["status"] == "approved"))
  end

  @doc "Most recent pending record for (stack, action), or nil."
  @spec latest_pending(String.t(), String.t(), String.t()) :: map() | nil
  def latest_pending(stack, project_id, action) do
    Enum.find(list_approvals(project_id), &(&1["stack"] == stack and &1["action"] == action and &1["status"] == "pending"))
  end

  @doc "Feature-flag skip gate (UC128): approval.skip.* / auto_approve keys."
  @spec should_skip_approval?(String.t(), String.t(), String.t(), keyword()) :: boolean()
  def should_skip_approval?(stack, project_id, action \\ "apply", opts \\ []) do
    env = Keyword.get(opts, :env, "prod") || "prod"
    org_id = Keyword.get(opts, :org_id)

    candidate_keys = [
      "approval.skip.#{action}",
      "approval.#{action}.skip",
      "approval.skip",
      "approval.auto_approve",
      "stack.#{stack}.skip_approval",
      "approval.stack.#{stack}.skip"
    ]

    Enum.any?(candidate_keys, fn key ->
      res = Flags.evaluate_scoped(key, env: env, project_id: project_id, org_id: org_id)
      res["enabled"] == true
    end)
  end

  def actions, do: @actions
end
